cogs/msg.py

```diff
@@ -37,13 +70,78 @@ async def set(self, ctx, name, *, content):
         content = await commands.clean_content().convert(ctx, content)
 
         with self.bot.db.get(ctx.guild.id) as db:
-            if len(db.execute("SELECT name, content FROM msg WHERE name = ?", (name.lower(),)).fetchall()) > 0:
-                db.execute("UPDATE msg SET content = ? WHERE name = ?", (content, name.lower()))
+            current_msg = db.execute(
+                "SELECT name, content, allow_memegen FROM msg WHERE name = ?", (name.lower(),)).fetchall()
+            if len(current_msg) > 0:
+                # only allow memegen if the current message is still an URL
+                keep_memegen = current_msg[0][2] and _is_url(content)
+                db.execute("UPDATE msg SET content = ?, allow_memegen = ? WHERE name = ?",
+                           (content, keep_memegen, name.lower()))
             else:
                 db.execute("INSERT INTO msg (name, content) VALUES (?, ?)", (name.lower(), content))
 
         await ctx.message.add_reaction('\U00002705')
 
+    @msg.command()
+    @basedbot.has_permissions("msg.set")
+    async def memegen(self, ctx, action, shorthand):
+        """Enable or disable meme generation for a shorthand"""
+
+        if action not in ("enable", "disable"):
+            await ctx.send("Invalid action. Valid actions are `enable` and `disable`")
+            return
+
+        with self.bot.db.get(ctx.guild.id) as db:
+            result = db.execute("SELECT name, content FROM msg WHERE name = ? OR name = ?",
+                                (shorthand.lower(), "-" + shorthand.lower())).fetchall()
+
+        if len(result) <= 0:
+            await ctx.send("Shorthand not found.")
+            return
+
+        shorthand_text = result[0][1]
+        if not _is_url(shorthand_text):
+            await ctx.send("Shorthand text must be a valid URL to manage meme generation.")
+            return
+
+        with self.bot.db.get(ctx.guild.id) as db:
+            result = db.execute("UPDATE msg SET allow_memegen = ? WHERE name = ? OR name = ?", (
+                action == "enable", shorthand.lower(), "-" + shorthand.lower()))
+
+        await ctx.message.add_reaction('\U00002705')
+
+    @msg.command()
+    @basedbot.has_permissions("msg.caption")
+    async def caption(self, ctx, shorthand, *, caption: commands.clean_content(
+        fix_channel_mentions=True, use_nicknames=True
+    )):
+        """Caption a meme"""
+
+        with self.bot.db.get(ctx.guild.id) as db:
+            result = db.execute("SELECT name, content, allow_memegen FROM msg WHERE name = ? OR name = ?",
+                                (shorthand.lower(), "-" + shorthand.lower())).fetchall()
+
+        if len(result) <= 0:
+            await ctx.send("Shorthand not found.")
+            return
+
+        if not result[0][2]:
+            await ctx.send("Meme generation has not been enabled for this shorthand.")
+            return
```

Comment:
Could probably just `SELECT` only entries where memegen has been enabled to begin with and adapt the "Shorthand not found" message accordingly ("... or captioning has not been enabled for the chosen shorthand").